### PR TITLE
Source sampling executable

### PR DESCRIPTION
**Pull request by DanShort12**
Adds an executable called `source_generator` to allow an `initial_source.h5` to be created without having to run a full OpenMC simulation.

This approach utilises the same OpenMC commands that would be run when running within the simulation. In particular:

*  `calculate_work`
*  `allocate_banks`
*  `initialize_source`

It also controls the settings needed to generate the initial source, allowing the user to specify the number of particles to sample, the output path, the verbosity, and the location of the source library (e.g. as built by the parametric plasma source). The run mode is fixed to eigenvalue, as this is the only OpenMC mode that generates the initial source output file, and the write initial source setting is fixed at true so that the output is always generated.

I've also included updates to the README to describe the build and run process.

### CODE REVIEW

**Review by makeclean — COMMENTED** [review]
Can you cmakeify that makefile?

**Comment by makeclean** [comment]
Sorry @DanShort12 can you rebase, then ill get this merged.

**Comment by DanShort12** [comment]
Rebased and cmakified.

File: README.md [hunk 1]
```diff
@@ -1,5 +1,68 @@
...
+
+# Usage
+
+The parametric plasma source can be imported an used in Python 3 in the following manner.
+
+```
+from parametric_plasma_source import Plasma
+my_plasma = Plasma(major_radius=6,
+                   minor_radius=1.5,
+                   elongation = 2.0
```
Comment (shimwell) [inline README.md]:
```suggestion
                   elongation = 2.0,
```

**Comment by shimwell** [comment]
Does anyone else get this error when trying ```python setup.py install```

```
  File "setup.py", line 8, in <module>
    from parametric_plasma_source.build_python import build as build_python
  File "/home/jshim/parametric-plasma-source/parametric_plasma_source/__init__.py", line 2, in <module>
    from .plasma_source import PlasmaSource
ModuleNotFoundError: No module named 'parametric_plasma_source.plasma_source'
```

File: README.md (hunk 1 above)
Comment (DanShort12) [reply README.md]: Thanks @Shimwell - I've fixed the typo and pushed to this PR.

**Comment by shimwell** [comment]
That appeared to help somewhat but not quite fully there with the setup.py, ```python setup.py install``` latest error ...
```
running develop
running egg_info
writing parametric_plasma_source.egg-info/PKG-INFO
writing dependency_links to parametric_plasma_source.egg-info/dependency_links.txt
writing top-level names to parametric_plasma_source.egg-info/top_level.txt
reading manifest file 'parametric_plasma_source.egg-info/SOURCES.txt'
writing manifest file 'parametric_plasma_source.egg-info/SOURCES.txt'
running build_ext
CMake Error at CMakeLists.txt:43 (add_subdirectory):
  The source directory

    /home/jshim/parametric-plasma-source/pybind11

  does not contain a CMakeLists.txt file.


CMake Error at CMakeLists.txt:45 (pybind11_add_module):
  Unknown CMake command "pybind11_add_module".


-- Configuring incomplete, errors occurred!
See also "/home/jshim/parametric-plasma-source/build/temp.linux-x86_64-3.7/CMakeFiles/CMakeOutput.log".
Traceback (most recent call last):
  File "setup.py", line 84, in <module>
    "Operating System :: OS Independent",
  File "/home/jshim/anaconda3/lib/python3.7/site-packages/setuptools/__init__.py", line 144, in setup
    return distutils.core.setup(**attrs)
  File "/home/jshim/anaconda3/lib/python3.7/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/home/jshim/anaconda3/lib/python3.7/distutils/dist.py", line 966, in run_commands
    self.run_command(cmd)
  File "/home/jshim/anaconda3/lib/python3.7/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "/home/jshim/anaconda3/lib/python3.7/site-packages/setuptools/command/develop.py", line 38, in run
    self.install_for_development()
  File "/home/jshim/anaconda3/lib/python3.7/site-packages/setuptools/command/develop.py", line 140, in install_for_development
    self.run_command('build_ext')
  File "/home/jshim/anaconda3/lib/python3.7/distutils/cmd.py", line 313, in run_command
    self.distribution.run_command(command)
  File "/home/jshim/anaconda3/lib/python3.7/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "setup.py", line 26, in run
    self.build_extension(ext)
  File "setup.py", line 56, in build_extension
    env=env
  File "/home/jshim/anaconda3/lib/python3.7/subprocess.py", line 363, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['cmake', '/home/jshim/parametric-plasma-source', '-DCMAKE_LIBRARY_OUTPUT_DIRECTORY=/home/jshim/parametric-plasma-source/parametric_plasma_source/', '-DPYTHON_EXECUTABLE=/home/jshim/anaconda3/bin/python', '-DCMAKE_BUILD_TYPE=Release']' returned non-zero exit status 1.
```

**Comment by DanShort12** [comment]
Couple of issues with building on a clean environment... all the more reason to get CI working! Thanks for trying it out @Shimwell - can you try again when you get a mo and let me know if it's working now?

**Comment by shimwell** [comment]
oh wow it looks like it works, that is a fancy setup.py it clones  git repos automatically and builds things.

**Comment by shimwell** [comment]
I shall continue to test, I'm not sure if this will need some more changes to get it working on the pypi repo but let me tinker for the day

**Comment by DanShort12** [comment]
Pretty sure it is likely to need a bit of tinkering for the PyPI repo but appreciate if you get a bit of time to try it out IRL (my poking has mostly been theoretical). I can rummage to try to find my script to read the initial_source.h5 into matplotlib scatter plots if that might help.

**Comment by DanShort12** [comment]
Oh, and the setup.py magic is in the interplay between the build and the CMakeLists.txt file - if you fancy a look through that file. The last bit of tinkering was to add some git commands into that CMake file.

Since the bindings are compiled, we'll need a bit of a think about how to package them for PyPI... they will be compile and Python version dependent.

**Comment by shimwell** [comment]
Perhaps this is just me but the ./source_generator --output is behaving slightly strangely

This command
``` ./source_generator -l test.so -o test.h5```
prints this to the terminal
```
 Number of particles: 1000
 Source library: test.so
 Output path: test.h5
 Verbosity: 5
```
but then makes a .h5 file called test.h5initial_source.h5

So it might be appending `initial_source.h5` to every output filename

**Comment by shimwell** [comment]
I've been tinkering and made some plots to show the plasma source with a few parameters chaining. I only tried the 4 shaping parameters (elongation, major_radius, minor_radius, triangularity) at the moment. I got stuck on shafranov_shift which appeared to have no impact on the plasma. But I need to check I'm doing everything correctly to be certain.


minor_radius 
triangularity


major_radius

elongation

**Comment by DanShort12** [comment]
Thanks @Shimwell, looks great! Just a suggestion, but the plots might be easier to interpret if they have some text in the corner to show what variable is changing and what the value being used is for a given frame (I know it's possible to deduce the variable from how the plot changes but that requires a little bit of bonus thought).

**Comment by makeclean** [comment]
Hmm shafranov shift should shift the equator radially outwards IIRC. Looking at the code as long as you’re using the sample source function it should work fine, if you short cut it and go some other route that avoids convert_rad_to_rz then you’ll not get the shift.

On 30 Jul 2020, at 17:23, Jonathan Shimwell <notifications@github.com<mailto:notifications@github.com>> wrote:



I've been tinkering and made some plots to show the plasma source with a few parameters chaining. I only tried the 4 shaping parameters (elongation, major_radius, minor_radius, triangularity) at the moment. I got stuck on shafranov_shift which appeared to have no impact on the plasma. But I need to check I'm doing everything correctly to be certain.

[minor_radius]<https://user-images.githubusercontent.com/8583900/88947809-fa99f380-d288-11ea-876a-735a4036b045.gif>
[triangularity]<https://user-images.githubusercontent.com/8583900/88947821-038ac500-d289-11ea-8614-7d3ee05f7130.gif>

[major_radius]<https://user-images.githubusercontent.com/8583900/88947788-f1a92200-d288-11ea-889d-1a1bfe968b65.gif>

[elongation]<https://user-images.githubusercontent.com/8583900/88947769-ea821400-d288-11ea-84a3-39fb8a98de07.gif>

**Comment by shimwell** [comment]
Congrats on the recent OpenMC merge 
https://github.com/openmc-dev/openmc/pull/1623

**Comment by DanShort12** [comment]
@Shimwell, this PR should now support passing the source parameters into the `source_generator` executable. Hope that's a useful addition!

**Comment by shimwell** [comment]
Sorry to be a pain is there any chance this PR can be redirect to the new repository

https://github.com/open-radiation-sources/parametric-plasma-source

@DanShort12

**Comment by DanShort12** [comment]
Superseded by open-radiation-sources/parametric-plasma-source#2

**Pull request closed without merging**